sdk/src/main/scala/com/horizen/account/state/AccountStateView.scala

```diff
@@ -103,44 +98,111 @@ class AccountStateView(private val metadataStorageView: AccountStateMetadataStor
-      }
-    } else
+  private def preCheck(tx: EthereumTransaction): BigInteger = {
+    // We are sure that transaction is semantically valid (so all the tx fields are valid)
+    // and was successfully verified by ChainIdBlockSemanticValidator
+
+    // Check signature
+    if (!tx.getSignature.isValid(tx.getFrom, tx.messageToSign()))
+      throw new TransactionSemanticValidityException(s"Transaction ${tx.id} is invalid: signature is invalid")
+
+    // Check that "from" is EOA address
+    if(!isEoaAccount(tx.getFrom.address()))
+      throw new TransactionSemanticValidityException(s"Transaction ${tx.id} is invalid: from account is not EOA")
+
+    // Check the nonce
+    val stateNonce: BigInteger = getNonce(tx.getFrom.address())
+    val txNonce: BigInteger = tx.getNonce
+    stateNonce.compareTo(txNonce) match {
+      case res if res > 0 =>
+        throw new TransactionSemanticValidityException(s"Transaction ${tx.id} is invalid: nonce ${txNonce} is to high (expected nonce is $stateNonce)")
+      case res if res < 0 =>
+        throw new TransactionSemanticValidityException(s"Transaction ${tx.id} is invalid: nonce ${txNonce} is to low (expected nonce is $stateNonce)")
+    }
+    if(txNonce.add(BigInteger.ONE).compareTo(txNonce) < 0)
+      throw new TransactionSemanticValidityException(s"Transaction ${tx.id} is invalid: nonce ${txNonce} reached the max value")
+
+    // Check eip15159 fee relation
+    if(tx.isEIP1559) {
+      // TODO:  tx.getMaxFeePerGas().compareTo(block base fee) < 0 -> exception: max fee per gas less than block base fee"
+    }
+
+    // Check that from account has enough balance to pay max gas*price value
+    // and reserve this amount.
+    val bookedGasPrice: BigInteger = buyGas(tx)
+
+    // Check that it is enough balance to pay after gas was bought.
+    val txBalanceAfterGasPrepayment: BigInteger = getBalance(tx.getFrom.address())
+    if(txBalanceAfterGasPrepayment.compareTo(tx.getValue) < 0)
+      throw new TransactionSemanticValidityException(s"Transaction ${tx.id} is invalid: not enough founds ${txBalanceAfterGasPrepayment} to pay ${tx.getValue}")
+
+    bookedGasPrice
+  }
+
+  private def buyGas(tx: EthereumTransaction): BigInteger = {
+    // TODO: implement gas prepayment strategy
+    // calc max possible payment
+    // subtract the balance
+    // return the used value
+    BigInteger.ZERO
+  }
+
+  override def applyTransaction(tx: SidechainTypes#SCAT): Try[Unit] = Try {
+    if (!tx.isInstanceOf[EthereumTransaction])
       throw new IllegalArgumentException(s"Unsupported transaction type ${tx.getClass.getName}")
+
+    val ethTx = tx.asInstanceOf[EthereumTransaction]
+
+    // Do the checks and prepay gas
+    val bookedGasPrice: BigInteger = preCheck(ethTx)
+
+    // Set Tx context for stateDB, to know where to keep EvmLogs
+    setupTxContext(ethTx)
+    val message: Message = Message.fromTransaction(ethTx)
+
+    // Increase the nonce by 1
+    increaseNonce(message.getFrom.address())
```

Comment:
Shall we increase the nonce here of at the very bottom of the method when we are sure we will not thrown any exception?

Reply:
we need to increase nonce in both success and fail.
That's why I put it before doing a snapshot

Reply:
right, in this case block is invalid and rejected.